File: period/c_backend.py
"""Experimental C backend for numeric/loop Period code.

Translates a subset of Period to C, compiles with the host C compiler,
and runs the resulting executable. Falls back to the interpreter for
unsupported constructs.
"""
import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from . import ast_nodes as ast


class CBackendError(Exception):
    """Raised when the C backend cannot handle a construct."""

    pass


_C_RUNTIME = r'''#include <stdio.h>
#include <stdint.h>

typedef int64_t period_int;
typedef double period_float;

static period_int period_abs(period_int x) { return x < 0 ? -x : x; }

static period_int _period_ipow(period_int base, period_int exp) {
    period_int result = 1;
    while (exp > 0) {
        if (exp & 1) result *= base;
        base *= base;
        exp >>= 1;
    }
    return result;
}
'''


class _CTranspiler:
    """Generate C source from a Period AST."""

    def __init__(self):
        self.functions: List[str] = []
        self.main_body: List[str] = []
        self._indent = "    "
        self._local_vars: set = set()

    def transpile(self, program: ast.Program) -> str:
        for stmt in program.statements:
            self._emit_top_level(stmt)
        return self._assemble()

    def _assemble(self) -> str:
        lines = [_C_RUNTIME]
        if self.functions:
            lines.extend(self.functions)
        lines.append("int main(void) {")
        lines.extend(self.main_body)
        lines.append("    return 0;")
        lines.append("}")
        return "\n".join(lines)

    def _emit_top_level(self, stmt: ast.Stmt):
        if isinstance(stmt, ast.DefineStmt):
            self.functions.append(self._transpile_function(stmt))
        elif isinstance(stmt, ast.ClassStmt):
            raise CBackendError("Classes are not supported by the C backend.")
        elif isinstance(stmt, ast.ImportStmt):
            raise CBackendError("Imports are not supported by the C backend.")
        else:
            self._emit_block_stmt(stmt, self.main_body, 1)

    def _transpile_function(self, stmt: ast.DefineStmt) -> str:
        params = ", ".join(f"period_int {p}" for p in stmt.parameters)
        body_lines: List[str] = []
        body_lines.append(f"static __forceinline period_int {stmt.name}({params}) {{")
        local_vars = set(stmt.parameters)
        for s in stmt.body:
            self._emit_block_stmt(s, body_lines, 1, local_vars)
        if not self._ends_with_return(stmt.body):
            body_lines.append("    return 0;")
        body_lines.append("}")
        return "\n".join(body_lines)

    def _ends_with_return(self, body: List[ast.Stmt]) -> bool:
        if not body:
            return False
        return isinstance(body[-1], ast.ReturnStmt)

    def _emit_block_stmt(
        self,
        stmt: ast.Stmt,
        out: List[str],
        level: int,
        local_vars: Optional[set] = None,
    ):
        if local_vars is None:
            local_vars = set()
        indent = self._indent * level
        if isinstance(stmt, ast.ExpressionStmt):
            expr = self._transpile_expr(stmt.expression)
            out.append(f"{indent}{expr};")
        elif isinstance(stmt, ast.LetStmt):
            expr = self._transpile_expr(stmt.initializer)
            out.append(f"{indent}period_int {stmt.name} = {expr};")
            local_vars.add(stmt.name)
        elif isinstance(stmt, ast.SetStmt):
            compound = self._try_compound_assignment(stmt.target, stmt.value)
            if compound:
                out.append(f"{indent}{compound};")
            else:
                target = self._transpile_expr(stmt.target)
                value = self._transpile_expr(stmt.value)
                out.append(f"{indent}{target} = {value};")
        elif isinstance(stmt, ast.ShowStmt):
            expr = self._transpile_expr(stmt.expression)
            out.append(f"{indent}printf(\"%lld\\n\", (long long){expr});")
        elif isinstance(stmt, ast.IfStmt):
            cond = self._transpile_expr(stmt.condition)
            out.append(f"{indent}if ({cond}) {{")
            for s in stmt.then_branch:
                self._emit_block_stmt(s, out, level + 1, local_vars)
            if stmt.else_branch:
                out.append(f"{indent}}} else {{")
                for s in stmt.else_branch:
                    self._emit_block_stmt(s, out, level + 1, local_vars)
            out.append(f"{indent}}}")
        elif isinstance(stmt, ast.WhileStmt):
            cond = self._transpile_expr(stmt.condition)
            out.append(f"{indent}while ({cond}) {{")
            for s in stmt.body:
                self._emit_block_stmt(s, out, level + 1, local_vars)
            out.append(f"{indent}}}")
        elif isinstance(stmt, ast.ForStmt):
            iterable = self._try_numeric_range(stmt.iterable)
            if iterable is None:
                raise CBackendError("Only 'range' iterables are supported by the C backend.")
            start, stop, step = iterable
            var = stmt.variable
            local_vars.add(var)
            out.append(f"{indent}for (period_int {var} = {start}; {var} < {stop}; {var} += {step}) {{")
            for s in stmt.body:
                self._emit_block_stmt(s, out, level + 1, local_vars)
            out.append(f"{indent}}}")
        elif isinstance(stmt, ast.ReturnStmt):
            if stmt.value is not None:
                expr = self._transpile_expr(stmt.value)
                out.append(f"{indent}return {expr};")
            else:
                out.append(f"{indent}return 0;")
        else:
            raise CBackendError(f"Statement type {type(stmt).__name__} is not supported by the C backend.")

    def _try_compound_assignment(self, target: ast.Expr, value: ast.Expr) -> Optional[str]:
        if not isinstance(target, ast.VariableExpr):
            return None
        if not isinstance(value, ast.BinaryExpr):
            return None
        if value.operator == "**":
            return None
        if not isinstance(value.left, ast.VariableExpr) or value.left.name != target.name:
            return None
        op_map = {"+": "+=", "-": "-=", "*": "*=", "/": "/=", "%": "%="}
        if value.operator not in op_map:
            return None
        right = self._transpile_expr(value.right)
        return f"{target.name} {op_map[value.operator]} {right}"

    def _try_numeric_range(self, expr: ast.Expr) -> Optional[Tuple[str, str, str]]:
        """If expr is a call to range with numeric args, return (start, stop, step) C expressions."""
        if not isinstance(expr, ast.CallExpr):
            return None
        callee = expr.callee
        if not isinstance(callee, ast.VariableExpr) or callee.name != "range":
            return None
        args = [self._transpile_expr(a) for a in expr.arguments]
        if len(args) == 1:
            return "0", args[0], "1"
        if len(args) == 2:
            return args[0], args[1], "1"
        if len(args) == 3:
            return args[0], args[1], args[2]
        return None

    def _transpile_expr(self, expr: ast.Expr) -> str:
        if isinstance(expr, ast.NumberLiteral):
            if isinstance(expr.value, float) and not expr.value.is_integer():
                return repr(expr.value)
            return repr(int(expr.value))
        if isinstance(expr, ast.BooleanLiteral):
            return "1" if expr.value else "0"
        if isinstance(expr, ast.NothingLiteral):
            return "0"
        if isinstance(expr, ast.VariableExpr):
            return expr.name
        if isinstance(expr, ast.BinaryExpr):
            left = self._transpile_expr(expr.left)
            right = self._transpile_expr(expr.right)
            op = expr.operator
            if op == "and":
                op = "&&"
            elif op == "or":
                op = "||"
            elif op == "==":
                op = "=="
            elif op == "!=":
                op = "!="
            elif op == "**":
                # Use integer exponentiation helper.
                return f"_period_ipow(({left}), ({right}))"
            return f"({left} {op} {right})"
        if isinstance(expr, ast.UnaryExpr):
            operand = self._transpile_expr(expr.operand)
            if expr.operator == "not":
                return f"(!{operand})"
            return f"({expr.operator}{operand})"
        if isinstance(expr, ast.CallExpr):
            callee = self._transpile_expr(expr.callee)
            args = ", ".join(self._transpile_expr(a) for a in expr.arguments)
            return f"{callee}({args})"
        if isinstance(expr, ast.IndexExpr):
            obj = self._transpile_expr(expr.object)
            idx = self._transpile_expr(expr.index)
            return f"({obj})[{idx}]"
        if isinstance(expr, ast.PropertyExpr):
            obj = self._transpile_expr(expr.object)
            return f"({obj}).{expr.name}"
        if isinstance(expr, ast.NewExpr):
            raise CBackendError("'new' expressions are not supported by the C backend.")
        if isinstance(expr, ast.TellExpr):
            raise CBackendError("'tell' expressions are not supported by the C backend.")
        if isinstance(expr, ast.QualifiedExpr):
            raise CBackendError("Module-qualified names are not supported by the C backend.")
        if isinstance(expr, ast.ListExpr):
            raise CBackendError("Lists are not supported by the C backend.")
        if isinstance(expr, ast.DictExpr):
            raise CBackendError("Dictionaries are not supported by the C backend.")
        if isinstance(expr, ast.StringLiteral):
            raise CBackendError("Strings are not supported by the C backend.")
        if isinstance(expr, ast.InputExpr):
            raise CBackendError("Input is not supported by the C backend.")
        raise CBackendError(f"Expression type {type(expr).__name__} is not supported by the C backend.")


def _find_windows_kits() -> Tuple[Optional[str], Optional[str]]:
    base = Path(r"C:\Program Files (x86)\Windows Kits\10")
    if not base.exists():
        base = Path(r"C:\Program Files\Windows Kits\10")
    if not base.exists():
        return None, None
    include_root = base / "Include"
    lib_root = base / "Lib"
    if not include_root.exists() or not lib_root.exists():
        return None, None
    versions = [d.name for d in include_root.iterdir() if d.is_dir()]
    if not versions:
        return None, None
    versions.sort(key=lambda s: tuple(int(x) for x in s.split(".")), reverse=True)
    version = versions[0]
    include = str(include_root / version)
    lib = str(lib_root / version)
    return include, lib


def find_cl() -> Optional[Path]:
    """Locate a usable MSVC cl.exe on Windows."""
    if sys.platform != "win32":
        return shutil.which("gcc") or shutil.which("clang") or None
    candidates = [
        Path(r"C:\Program Files (x86)\Microsoft Visual Studio"),
        Path(r"C:\Program Files\Microsoft Visual Studio"),
    ]
    for root in candidates:
        if not root.exists():
            continue
        for cl in root.rglob(r"VC\Tools\MSVC\*\bin\Hostx64\x64\cl.exe"):
            return cl
    return None


def compile_c(c_source: str, output_exe: Path) -> Tuple[bool, str]:
    """Compile generated C source to an executable."""
    cl = find_cl()
    if cl is None:
        return False, "No C compiler found. Install MSVC, gcc, or clang."

    with tempfile.NamedTemporaryFile("w", suffix=".c", delete=False) as f:
        f.write(c_source)
        c_path = Path(f.name)

    try:
        if sys.platform == "win32" and "cl.exe" in str(cl).lower():
            include_dir = (cl.parents[3] / "include").as_posix()
            lib_dir = (cl.parents[3] / "lib" / "x64").as_posix()
            win_include, win_lib = _find_windows_kits()
            if not win_include or not win_lib:
                return False, "Windows SDK not found."
            cmd = [
                str(cl),
                "/O2",
                f"/I{include_dir}",
                f"/I{win_include}/ucrt",
                f"/I{win_include}/um",
                f"/I{win_include}/shared",
                f"/Fe:{output_exe.as_posix()}",
                c_path.as_posix(),
                "/link",
                f"/LIBPATH:{lib_dir}",
                f"/LIBPATH:{win_lib}/ucrt/x64",
                f"/LIBPATH:{win_lib}/um/x64",
            ]
            env = os.environ.copy()
            env["MSYS2_ARG_CONV_EXCL"] = "*"
            proc = subprocess.run(cmd, capture_output=True, text=True, env=env)
        else:
            cmd = [str(cl), "-O2", "-o", str(output_exe), str(c_path)]
            proc = subprocess.run(cmd, capture_output=True, text=True)

        if proc.returncode != 0:
            return False, proc.stdout + proc.stderr
        return True, ""
    finally:
        c_path.unlink(missing_ok=True)


def run_native(program: ast.Program) -> Tuple[bool, str, str]:
    """Transpile, compile, and run the program. Returns (ok, stdout, stderr)."""
    transpiler = _CTranspiler()
    try:
        c_source = transpiler.transpile(program)
    except CBackendError as exc:
        return False, "", str(exc)

    with tempfile.TemporaryDirectory() as tmpdir:
        exe = Path(tmpdir) / "period_native.exe"
        ok, err = compile_c(c_source, exe)
        if not ok:
            return False, "", err
        proc = subprocess.run([str(exe)], capture_output=True, text=True)
        return True, proc.stdout, proc.stderr

File: period/src/c_backend.rs
use std::collections::HashSet;
use crate::ast::*;

pub fn try_compile_c(program: &Program) -> Option<String> {
    let mut generator = CGen::new();
    if generator.gen_program(program).is_ok() {
        Some(generator.output)
    } else {
        None
    }
}

#[derive(Debug)]
struct Unsupported;

struct CGen {
    output: String,
    indent: usize,
    locals: Vec<HashSet<String>>,
    globals: HashSet<String>,
}

impl CGen {
    fn new() -> Self {
        Self { output: String::new(), indent: 0, locals: Vec::new(), globals: HashSet::new() }
    }

    fn unsupported<T>() -> Result<T, Unsupported> { Err(Unsupported) }

    fn in_function(&self) -> bool { !self.locals.is_empty() }

    fn current_locals(&self) -> &HashSet<String> {
        self.locals.last().unwrap_or(&self.globals)
    }

    fn is_local(&self, name: &str) -> bool {
        self.locals.iter().any(|scope| scope.contains(name))
    }

    fn line(&mut self, s: &str) {
        for _ in 0..self.indent { self.output.push_str("    "); }
        self.output.push_str(s);
        self.output.push('\n');
    }

    fn gen_program(&mut self, program: &Program) -> Result<(), Unsupported> {
        self.line("#include <stdio.h>");
        self.line("");
        self.line("static long long period_pow(long long base, long long exp) {");
        self.indent += 1;
        self.line("long long result = 1;");
        self.line("while (exp > 0) { result *= base; exp--; }");
        self.line("return result;");
        self.indent -= 1;
        self.line("}");
        self.line("");

        // Emit top-level functions first.
        for stmt in &program.statements {
            if let Stmt::Define { .. } = stmt {
                self.gen_toplevel_stmt(stmt)?;
            }
        }

        // Wrap top-level statements in main().
        self.line("int main(void) {");
        self.indent += 1;
        for stmt in &program.statements {
            if !matches!(stmt, Stmt::Define { .. }) {
                self.gen_stmt(stmt)?;
            }
        }
        self.line("return 0;");
        self.indent -= 1;
        self.line("}");
        Ok(())
    }

    fn gen_toplevel_stmt(&mut self, stmt: &Stmt) -> Result<(), Unsupported> {
        match stmt {
            Stmt::Define { name, params, body, .. } => self.gen_define(name, params, body),
            _ => self.gen_stmt(stmt),
        }
    }

    fn gen_define(&mut self, name: &str, params: &[(String, Option<String>)], body: &[Stmt]) -> Result<(), Unsupported> {
        let params_str = params.iter().map(|(n, _)| format!("long long {}", n)).collect::<Vec<_>>().join(", ");
        self.line(&format!("static long long {}({}) {{", name, params_str));
        self.indent += 1;
        let mut locals: HashSet<String> = params.iter().map(|(n, _)| n.clone()).collect();
        for stmt in body {
            Self::collect_locals(stmt, &mut locals);
        }
        self.locals.push(locals);
        for stmt in body { self.gen_stmt(stmt)?; }
        self.line("return 0;");
        self.indent -= 1;
        self.line("}");
        self.locals.pop();
        Ok(())
    }

    fn collect_locals(stmt: &Stmt, locals: &mut HashSet<String>) {
        match stmt {
            Stmt::Let { name, .. } => { locals.insert(name.clone()); }
            Stmt::If { then_branch, else_branch, .. } => {
                for s in then_branch { Self::collect_locals(s, locals); }
                for s in else_branch { Self::collect_locals(s, locals); }
            }
            Stmt::While { body, .. } | Stmt::For { body, .. } => {
                for s in body { Self::collect_locals(s, locals); }
            }
            _ => {}
        }
    }

    fn gen_stmt(&mut self, stmt: &Stmt) -> Result<(), Unsupported> {
        match stmt {
            Stmt::Let { name, value } => {
                self.globals.insert(name.clone());
                let val = self.gen_expr(value)?;
                self.line(&format!("long long {} = {};", name, val));
            }
            Stmt::Set { target, value } => {
                let target_str = self.gen_assign_target(target)?;
                let val = self.gen_expr(value)?;
                self.line(&format!("{} = {};", target_str, val));
            }
            Stmt::Show(expr) => {
                let val = self.gen_expr(expr)?;
                self.line(&format!("printf(\"%lld\\n\", {});", val));
            }
            Stmt::If { cond, then_branch, else_branch } => {
                let c = self.gen_cond(cond)?;
                self.line(&format!("if ({}) {{", c));
                self.indent += 1;
                for s in then_branch { self.gen_stmt(s)?; }
                self.indent -= 1;
                if else_branch.is_empty() {
                    self.line("}");
                } else {
                    self.line("} else {");
                    self.indent += 1;
                    for s in else_branch { self.gen_stmt(s)?; }
                    self.indent -= 1;
                    self.line("}");
                }
            }
            Stmt::While { cond, body } => {
                let c = self.gen_cond(cond)?;
                self.line(&format!("while ({}) {{", c));
                self.indent += 1;
                for s in body { self.gen_stmt(s)?; }
                self.indent -= 1;
                self.line("}");
            }
            Stmt::For { var, iterable, body } => {
                let iter = self.gen_iterable(iterable)?;
                self.line(&format!("for (long long {var} = 0; {var} < {iter}; {var}++) {{"));
                self.indent += 1;
                for s in body { self.gen_stmt(s)?; }
                self.indent -= 1;
                self.line("}");
            }
            Stmt::Return(Some(expr)) => {
                let val = self.gen_expr(expr)?;
                self.line(&format!("return {};", val));
            }
            Stmt::Return(None) => self.line("return 0;"),
            Stmt::Expr(expr) => { self.gen_expr(expr)?; }
            _ => return Self::unsupported(),
        }
        Ok(())
    }

    fn gen_assign_target(&mut self, target: &AssignTarget) -> Result<String, Unsupported> {
        match target {
            AssignTarget::Variable(name) => Ok(name.clone()),
            AssignTarget::Index { .. } => Self::unsupported(),
            AssignTarget::Property { .. } => Self::unsupported(),
        }
    }

    fn gen_iterable(&mut self, expr: &Expr) -> Result<String, Unsupported> {
        match expr {
            Expr::Call { callee, args } => {
                if let Expr::Variable { name, .. } = callee.as_ref() {
                    if name == "range" {
                        let args_str: Result<Vec<_>, _> = args.iter().map(|a| self.gen_expr(a)).collect();
                        return Ok(args_str?.join(""));
                    }
                }
                Self::unsupported()
            }
            _ => Self::unsupported(),
        }
    }

    fn gen_cond(&mut self, expr: &Expr) -> Result<String, Unsupported> {
        // Conditions in C can use raw comparisons (non-zero == true).
        match expr {
            Expr::Bool(b) => Ok(if *b { "1" } else { "0" }.to_string()),
            Expr::Number(n) => Ok(format!("{}LL", *n as i64)),
            Expr::Variable { name, .. } => Ok(name.clone()),
            Expr::Binary { op, left, right } => {
                let l = self.gen_expr(left)?;
                let r = self.gen_expr(right)?;
                match op {
                    BinOp::Eq => Ok(format!("({}) == ({})", l, r)),
                    BinOp::Ne => Ok(format!("({}) != ({})", l, r)),
                    BinOp::Lt => Ok(format!("({}) < ({})", l, r)),
                    BinOp::Gt => Ok(format!("({}) > ({})", l, r)),
                    BinOp::Le => Ok(format!("({}) <= ({})", l, r)),
                    BinOp::Ge => Ok(format!("({}) >= ({})", l, r)),
                    BinOp::And => Ok(format!("({}) && ({})", l, r)),
                    BinOp::Or => Ok(format!("({}) || ({})", l, r)),
                    _ => Ok(format!("({}) != 0", self.gen_expr(expr)?)),
                }
            }
            Expr::Unary { op, operand } => {
                let v = self.gen_expr(operand)?;
                match op {
                    UnaryOp::Neg => Ok(format!("(-{}) != 0", v)),
                    UnaryOp::Not => Ok(format!("!{}", v)),
                }
            }
            _ => Ok(format!("({}) != 0", self.gen_expr(expr)?)),
        }
    }

    fn gen_expr(&mut self, expr: &Expr) -> Result<String, Unsupported> {
        match expr {
            Expr::Number(n) => Ok(format!("{}LL", *n as i64)),
            Expr::Bool(b) => Ok(if *b { "1LL" } else { "0LL" }.to_string()),
            Expr::Variable { name, .. } => Ok(name.clone()),
            Expr::Binary { op, left, right } => {
                let l = self.gen_expr(left)?;
                let r = self.gen_expr(right)?;
                match op {
                    BinOp::Add => Ok(format!("({} + {})", l, r)),
                    BinOp::Sub => Ok(format!("({} - {})", l, r)),
                    BinOp::Mul => Ok(format!("({} * {})", l, r)),
                    BinOp::Div => Ok(format!("({} / {})", l, r)),
                    BinOp::Mod => Ok(format!("({} % {})", l, r)),
                    BinOp::Pow => Ok(format!("period_pow({}, {})", l, r)),
                    BinOp::Eq => Ok(format!("(({}) == ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Ne => Ok(format!("(({}) != ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Lt => Ok(format!("(({}) < ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Gt => Ok(format!("(({}) > ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Le => Ok(format!("(({}) <= ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Ge => Ok(format!("(({}) >= ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::And => Ok(format!("(({}) && ({}) ? 1LL : 0LL)", l, r)),
                    BinOp::Or => Ok(format!("(({}) || ({}) ? 1LL : 0LL)", l, r)),
                }
            }
            Expr::Unary { op, operand } => {
                let v = self.gen_expr(operand)?;
                match op {
                    UnaryOp::Neg => Ok(format!("(-{})", v)),
                    UnaryOp::Not => Ok(format!("(!{} ? 1LL : 0LL)", v)),
                }
            }
            Expr::Call { callee, args } => {
                let name = match callee.as_ref() {
                    Expr::Variable { name, .. } => name.clone(),
                    _ => return Self::unsupported(),
                };
                let args_str: Result<Vec<_>, _> = args.iter().map(|a| self.gen_expr(a)).collect();
                Ok(format!("{}({})", name, args_str?.join(", ")))
            }
            _ => Self::unsupported(),
        }
    }
}
